### ISSUE TITLE
`Vec{N, T}` constructor should construct `Vec{N, T}` instance

### ISSUE THREAD

**Issue by hyrodium** [JuliaGeometry/Meshes.jl#758]
MRE:
```julia
julia> using Meshes

julia> Vec{3,Int}(1,2,3)
Vec(1.0, 2.0, 3.0)

julia> Vec{3,Int}(1,2,3) isa Vec{3,Int}
false
```

This causes the following error with broadcasting.

```julia
julia> using Meshes, StaticArrays

julia> string.(SVector(1,2,3))
3-element SVector{3, String} with indices SOneTo(3):
 "1"
 "2"
 "3"

julia> string.(Vec(1,2,3))
ERROR: MethodError: no method matching AbstractFloat(::Type{String})

Closest candidates are:
  AbstractFloat(::Int8)
   @ Base float.jl:265
  AbstractFloat(::Int16)
   @ Base float.jl:266
  AbstractFloat(::Bool)
   @ Base float.jl:264
  ...
```

These behaviors are based on the following method definition with `float`.

https://github.com/JuliaGeometry/Meshes.jl/blob/16510ca92386c5632167b35d38d51d75f7dbd3dc/src/vectors.jl#L44-L56

If you agree that accepting `Vec` with non-float elements is okay, I will open a PR!

**Comment by juliohm** [comment]
We opted for a continuous representation of coordinates, and designed these constructors to never allow discrete spaces (e.g. integers). Can you elaborate on why you need integer coordinates?

**Comment by hyrodium** [comment]
>Can you elaborate on why you need integer coordinates?

I actually don't have a practical application with integer coordinates. That was just a minimum example.

As I said in the title, the problem is the return type of the constructor. Most users *definitely* expect return type `T` if the users explicitly called `T(x)`. (cf. [related official documentation about parametric constructors](https://docs.julialang.org/en/v1/manual/constructors/#Parametric-Constructors))

---

We have two choices to resolve this issue.

* Make `Vec{3,Int}(1,2,3)` throws an error.
* Remove the restriction with floating point numbers.

## Make `Vec{3,Int}(1,2,3)` throws an error.

This is the easiest way to resolve the issue. Just updating the L50-56 with:

```julia
 function Vec{Dim,T}(coords::Union{Tuple,AbstractVector}) where {Dim,T<:AbstractFloat}
   if Dim ≠ length(coords) 
     throw(DimensionMismatch("invalid dimension")) 
   end 
   Vec(NTuple{Dim,T}(coords)) 
 end 
```

## Remove the restriction with floating point numbers.

I prefer this solution to the previous one.

* I initially called something like `string.(Vec(1,2,3))` to export the values to a text file. Relaxing the constructor can avoid this kind of error.
* Vectors with integer elements are useful for [Pick's theorem](https://en.wikipedia.org/wiki/Pick%27s_theorem). I think supporting `Vec{N, Int}` is not that bad idea.
* There was a comment about the element type on discourse. (https://discourse.julialang.org/t/ann-announcing-meshes-jl/53973/38)
* Someone may want to use `FixedPointNumbers.Fixed` to represent the elements of a vector. This type is an example of a continuous representation of coordinates without floating point numbers.

**Comment by juliohm** [comment]
> As I said in the title, the problem is the return type of the constructor. Most users _definitely_ expect return type `T` if the users explicitly called `T(x)`. (cf. [related official documentation about parametric constructors](https://docs.julialang.org/en/v1/manual/constructors/#Parametric-Constructors))

On the other hand, we had multiple reports of users having trouble with geometric algorithms because they fed geometries with Integer coordinates. The algorithms assume continuous coordinates, and the errors were cryptic to end-users. That was a design decision that we made a long time ago to make sure that code written as `Vec(1,2,3)` still works.

This parameter `T` will be revisited soon anyways given the new CRS infrastructure.

**Comment by hyrodium** [comment]
>That was a design decision that we made a long time ago to make sure that code written as `Vec(1,2,3)` still works.

My point is, `Vec{3, Int}(1,2,3)` should be `Vec{3, Int}` or throw an error. `Vec(1,2,3) isa Vec{3, Float64}` is not problematic at all.

I opened a related PR https://github.com/JuliaLang/julia/pull/53328 that updates the style guide.

**Comment by juliohm** [comment]
I see your point now. It makes total sense to respect the type in that case. Appreciate if you can submit a PR that throws an error.

**Comment by eliascarv** [comment]
Resolved by #763

**Comment by hyrodium** [comment]
Thank you for the fix!
